### PR TITLE
Depth map refactoring

### PR DESCRIPTION
**Pull request by fabiencastan**
## Description

Use more modern cuda structures and remove all global variables.

## Features list

- [X] Remove dead code
- [X] Replace all global variables with explicit arguments
- [X] SGM: keep all similarity volumes only on the device
- [X] Reduce number of buffers in SGM filtering
- [X] Remove buffer copy/inversion in SGM filtering
- [X] Build option to use uchar or float image buffers
- [X] Use dedicated buffer for image variance (instead of alpha channel
- [X] Use alpha channel from input images to filter part of the images

Also include  #715

## Implementation remarks

Same performances than the develop branch for now.
Need to validate quality impact.

### CODE REVIEW

File: src/aliceVision/depthMap/cuda/planeSweeping/plane_sweeping_cuda.cu [hunk 1]
```diff
@@ -961,20 +957,40 @@ void ps_optimizeDepthSimMapGradientDescent(Pyramids& ps_texs_arr,
             optDepthMap_dmp.getBuffer(), optDepthMap_dmp.getPitch(),
             optDepthSimMap_dmp.getBuffer(), optDepthSimMap_dmp.getPitch(),
             width, height);
-        copy(optDepthMap_arr, optDepthMap_dmp);
-        // Bind those depth values as a texture
-        cudaBindTextureToArray(depthsTex, optDepthMap_arr.getArray(), cudaCreateChannelDesc<float>());
 
-        // Adjust depth/sim by using previously computed depths (depthTex is accessed inside this kernel)
```
Comment (fabiencastan) [inline src/aliceVision/depthMap/cuda/planeSweeping/plane_sweeping_cuda.cu]: @griwodz Can you review this commit?

**Comment by marcfielding1** [comment]
Hey guys awesome work, any rough idea when this will be in develop, I dying to try this out.

**Comment by natowi** [comment]
@marcfielding1  [With the merge of this PR](https://github.com/alicevision/AliceVision/pull/753) you will be able to download a pre-build snapshot via appveyor.

**Comment by marcfielding1** [comment]
@natowi  Heya sorry I forgot to reply, seems the appveyor build bugged out on this - apologies for being so slow to reply this got lost in my notifications!

**Comment by natowi** [comment]
@marcfielding1 yes, sorry, the appveyor auto-build support is sadly broken and will be replaced by a github build action, but so far it has not been implemented

**Comment by marcfielding1** [comment]
@natowi Heya, thanks for the reply, so a) is this merged/can I try it out b) I might be able to sort the actions integration - no promises as to time but do you have a set of requirements for replacing appveyor so I can estimate?

**Comment by natowi** [comment]
@marcfielding1 That would be great and appreciated. The Github Action for Linux CI has been merged https://github.com/alicevision/AliceVision/pull/843 some time ago, getting windows CI would be great - especially with auto-builds. Best get in contact with @fabiencastan for details on the requirements.

**Comment by marcfielding1** [comment]
@natowi yeah I'd love to get involved - my company is intending on using this in production, we've created photogrammetry technology that sits in a photobooth for retailers to use across FMCG product lines, with this feature and other upcoming ones it allows us to continue to use the software without perhaps moving to paid alternatives.

**Comment by marcfielding1** [comment]
@fabiencastan / @natowi Sorry I just reread the comments from Feb, we really really wanna try out image masking, we're using a photo booth with a baseplate to fool PG into thinking it's moving around the object when actually the object is moving - #715 is very interesting for this purpose to cut out background noise.

Can I just yank this branch and build it locally? 

Secondly, what do you need in terms of auto-build stuff? I'll see if I can find time to do it!

Cheers

Marc

**Comment by natowi** [comment]
You could take a look at the work by EwoutH https://github.com/alicevision/AliceVision/pull/638#issuecomment-622957884 and maybe help to finalize the PR

**Review by gregoire-dl — APPROVED** [review]

**Pull request merged**